155_shiftapp_sprint22_issue155_employe-time-page

MiguelQuintana1 opened the pull request:
… integration, with the capasity of being tested by user role

<!-- Please fill out the following: -->
## Summary & Changes 📃
- **Resolves:** #155 

- **Summary:** (Briefly describe what this PR does)
  
  ## Component Explanation: EmployeeTimeScreen

The `EmployeeTimeScreen` is a React Native component built using Expo that displays employee time information, allowing a user (either a supervisor or an individual employee) to view scheduled and maximum work hours. It fetches and presents this data in a list format, with a toggle to control the visibility of detailed scheduled hours.

**Core Functionality:**

1.  **Role-Based Data Display:** The component simulates displaying data based on the user's role (`supervisor` or `employee`). A `userRole` state allows for testing this behavior.
    * **Supervisor:** When the `userRole` is 'supervisor', the component fetches and displays data for multiple employees. Each employee's information is shown in a separate card.
    * **Employee:** When the `userRole` is 'employee', the component fetches and displays data for a single employee (representing the logged-in user).

2.  **Independent Toggle for Scheduled Hours (Supervisor View):** For supervisors, each employee card has its own "Show Scheduled Hours" toggle (`Switch`). This allows the supervisor to independently view or hide the detailed scheduled hours (represented by a `TimeLineBar`) for each employee.

3.  **Timeline Visualization:** When the "Show Scheduled Hours" toggle is on, a `TimeLineBar` component is rendered. This bar visually represents the proportion of scheduled hours to the maximum allowed hours for an employee, using different colors for scheduled and remaining time.

4.  **Loading and Error Handling:** The component includes basic loading and error states (`loading` and `error`). While data is being fetched, a "Loading..." message is displayed. If an error occurs during data fetching, an error message is shown.

5.  **Data Fetching Simulation:** The `fetchEmployeeData` function simulates fetching data from a backend API. It uses `setTimeout` to mimic network latency and provides mock data based on the current `userRole`. In a real application, this function would be replaced with an actual API call.

**Component Breakdown:**

* **`EmployeeTimeScreen` Functional Component:**
    * **State Variables:**
        * `employeeData`: An array of `Employee` objects fetched and displayed by the component.
        * `loading`: A boolean indicating if data is currently being fetched.
        * `error`: A string containing an error message if data fetching fails (null otherwise).
        * `userRole`: A string (`'supervisor'` or `'employee'`) to simulate different user roles for testing.
    * **`fetchEmployeeData` Asynchronous Function:** Simulates fetching employee data based on the `userRole`. It updates the `employeeData`, `loading`, and `error` states.
    * **`useEffect` Hook:** Calls `fetchEmployeeData` when the component mounts and whenever the `userRole` changes, ensuring the data is updated based on the current role.
    * **`handleToggleSwitch` Function:** Takes an `employeeId` and updates the `showScheduled` property of the corresponding employee in the `employeeData` state, enabling the independent toggle functionality for supervisors.
    * **`renderEmployeeCard` Function:** Renders a card for each employee in the `employeeData` array. This card displays the employee's name, conditionally renders the `TimeLineBar` or the "Max Hours" text based on the `showScheduled` state, and includes the "Show Scheduled Hours" toggle.
    * **Return Statement:** Renders the main view, including a header, a role switching mechanism (for testing), a loading indicator, an error message (if any), and the list of employee cards using `FlatList`.

* **`TimeLineBar` Functional Component:**
    * **Props:** Takes `scheduled` (number) and `max` (number) as props, representing the scheduled and maximum hours for an employee.
    * **Functionality:** Calculates the remaining hours and the percentages for scheduled and remaining time. It then renders a visual timeline bar using nested `View` components with dynamic widths based on these percentages. It also displays the numerical representation of scheduled and maximum hours.

* **Type Definitions:**
    * `Employee`: Defines the structure of an employee object, including `id`, `name`, `scheduledHours`, `maxHours`, and `showScheduled` (for the toggle state).
    * `TimeLineBarProps`: Defines the expected props for the `TimeLineBar` component.

* **Styling (`StyleSheet`):** The `styles` object defines the visual appearance of the various components and elements within the `EmployeeTimeScreen`, including containers, headers, lists, cards, text, toggles, and the timeline bar.

**In summary, this component provides a user interface for viewing employee time information, adapting its display based on the user's role and offering an interactive, per-employee toggle for detailed scheduled hour visualization in a supervisor context.**


to test this component Check the video:

https://github.com/user-attachments/assets/d61308f1-4434-4b3c-9bdd-21b40a498fdc





## How to Test 🧪
1-  npm install
2- supabase start
3- npx expo start
4- in the main page go to Employee Time Page
5- to test depending on the role. in your local enviroment project go to "components" folder -> EmployeeTimePage.tsx.
in that component line 47 you can change the role from "supervisor" to "employee".


## Checklist ✅

- [ ] I have **tested** this PR **locally** and it works as expected.
- [ ] This PR **resolves an issue** (`Resolves #issue-number`).
- [ ] **Reviewers, assignees(self), tags, and labels** are correctly assigned. <!-- on the menu to the right -->
- [ ] **Squash commits** and enable **auto-merge** if approved.

MiguelQuintana1 committed:
creation of the component EmployeeTimePage, thinking in a future data integration, with the capasity of being tested by user role

el-riber reviewed (commented):
When I went I tried to run the supabase start command, it failed. Can someone help me with this:
![Screenshot (145)](https://github.com/user-attachments/assets/a955dc6c-d7fd-4d01-9bd2-0072cf7f12be)

cshimm commented:
@el-riber I believe the command you are looking for is `npx supabase start`!

cshimm reviewed (approved):
Nice job, works as expected on my end!
![image](https://github.com/user-attachments/assets/5057e42f-b390-4159-b2f9-c9cf0caedcfa)

IsaacJrTypes reviewed (dismissed):
Good job on the employee time limit page component. I have a few suggestions that I think could be easily implemented:
- Could to add color status depending on how close the employee is to their max alloted hours? Ex) If the user meets or exceeds their max, the progress bar turns red
- Remove my-expo-app directory
- Unsure why we are creating a toggle to show or hide the progress bar, but feel free to provide a convincing perspective. If we are going to hide the progress bar, supervisors should be able to quickly tell if users have hours to pick up shifts. You could update the component to include the number of hours worked to max hours or remove the toggle functionality completely.

I would place the number of hours worked next to max hours or use rations to indicate progress. Ex) 5h / 20h Max hours worked 
![CleanShot 2025-05-12 at 22 08 05](https://github.com/user-attachments/assets/2089f4b2-1ef9-4be4-b6b0-e0d1c1fa6fe3)

MiguelQuintana1 committed:
added new functionalities, deleted the toggle

MiguelQuintana1 commented:
https://github.com/user-attachments/assets/7f101318-310a-4ac5-9371-f3a2711e1896

Summary of Changes Made to the EmployeeTimePage Component:
Added Remaining Time Display:

Added a new section under the progress bar to display the "Remaining Time Available".
The remaining time is calculated as [max - scheduled](vscode-file://vscode-app/private/var/folders/tq/sgp04zpd3rggmwrv6t7lc1bw0000gn/T/AppTranslocation/CB57F5E1-F2C7-4FBA-B9DB-D06D986EE268/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) and displayed in the format:
Remaining Time Available: Xh.
Dynamic Progress Bar Color:

Implemented logic to dynamically change the progress bar's color based on the percentage of scheduled hours:
Green: Less than 50% of max hours.
Yellow: Between 50% and 79% of max hours.
Red: 80% or more of max hours.
Separated Views for Clarity:

Placed the "Remaining Time Available" text in a separate [View](vscode-file://vscode-app/private/var/folders/tq/sgp04zpd3rggmwrv6t7lc1bw0000gn/T/AppTranslocation/CB57F5E1-F2C7-4FBA-B9DB-D06D986EE268/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) below the progress bar for better visual separation.
Updated Styles:

Added a new style [remainingTimeContainer](vscode-file://vscode-app/private/var/folders/tq/sgp04zpd3rggmwrv6t7lc1bw0000gn/T/AppTranslocation/CB57F5E1-F2C7-4FBA-B9DB-D06D986EE268/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) to create spacing for the Remaining Time section.
Added a new style [remainingTime](vscode-file://vscode-app/private/var/folders/tq/sgp04zpd3rggmwrv6t7lc1bw0000gn/T/AppTranslocation/CB57F5E1-F2C7-4FBA-B9DB-D06D986EE268/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) to format the Remaining Time Available text.
Mock Data Adjustments:

Updated mock data in the [fetchEmployeeData](vscode-file://vscode-app/private/var/folders/tq/sgp04zpd3rggmwrv6t7lc1bw0000gn/T/AppTranslocation/CB57F5E1-F2C7-4FBA-B9DB-D06D986EE268/d/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html) function to test the new features:
Example: Jane Doe has a max of 20 hours and is scheduled for 8 hours, which is 40% of her time.
Preserved Existing Features:

Kept the existing format for the progress bar label (Xh / Yh Max).
Ensured the component works seamlessly for both supervisor and employee roles.

Toggle removed completely.

Thank you Issac For the suggestions.

MiguelQuintana1 committed:
not sure

IsaacJrTypes reviewed (approved):
The components looks much better. Thank you for making the changes!
![image](https://github.com/user-attachments/assets/516b91a6-cd19-4044-92e0-614c50df9352)
